### ISSUE TITLE
Code cell shows "markdown" language

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#104606]
I can't repro this manually but I can hit it with a test reliably. Somehow we

- insert a code cell
- send the splice to the EH, it's handled in `_spliceNotebookCells`
- Call `this._documentsAndEditors.acceptDocumentsAndEditorsDelta` with the addedDocuments
- `let ref = this._documents.get(resource);`
- There is already a cell that exists with that URI, and it is a markdown cell

![image](https://user-images.githubusercontent.com/323878/90202199-e4645b00-dd91-11ea-9e63-6d6148f9d07c.png)

- So `ExtHostCell#language` for this new cell returns `markdown`, which is incorrect

**Comment by roblourens** [comment]
I triggered it manually but I'm not sure how:

![image](https://user-images.githubusercontent.com/323878/90202833-b08a3500-dd93-11ea-98cb-9d129c27573d.png)

**Comment by roblourens** [comment]
I'm done debugging for tonight but I suspect that the issue involves editing a notebook while it is closed (like outside of vscode). The test opens a notebook with the same URI but different content, and this only fails on the second run. These steps seem related:

- Open a github issue notebook
- Close it
- Open it in vim, delete some cells
- Open it in vscode, insert some cells
- Save it
- Sometimes those cells are saved with the wrong content

**Comment by rebornix** [comment]
@roblourens in the smash test, can you add a validation after a document is opened in each test, and check if the document is the same as what's on disk? If it doesn't match, it means the document is obsolete and definitely a bug with our document lifecycle.

**Comment by roblourens** [comment]
I can't repro an issue with my manual steps above, but the test still hit this reliably and I think there is a real issue that I don't fully understand. @jrieken, help my understand what's going on. I added logging to https://github.com/Microsoft/vscode/blob/roblou/notebookRandomActionTest/src/vs/workbench/api/common/extHostDocumentsAndEditors.ts#L80

and it shows that every cell gets "added" twice and "removed" twice, so the ref counting cleans it up nicely. But, in my test case, one cell only gets "removed" once so when it is added again on the second test run, an entry with that URI already exists. Why are cells ref'd twice and where should I look to figure out why one of them is not unref'd?

**Comment by jrieken** [comment]
> Why are cells ref'd twice and where should I look to figure out why one of them is not unref'd?

In the UI, when a notebook opens we don't create text models because that would be slow. However, text models drive the TextDocument api and we want each cell to appear as text document in the API as soon as a notebook is opened. Therefore, we open each cell as text document (this happens in [the extension host](https://github.com/microsoft/vscode/blob/442e27637de2ec4b24337299e49a1bb17071b9f5/src/vs/workbench/api/common/extHostNotebook.ts#L407-L409)) when a notebook opens, later the UI might also open these cell documents which ends up running the same code path and increasing the ref counting. When notebook is closed in dereferences its documents, similar things happen in the renderer - when a text model gets disposed it dereferences the extension host document. You can have a dangling reference when one of those things doesn't happen. Since references on the UI side are also ref-counted it is likely that there are dangling references, which can happen when using the `openTextDocument`-API which (on the UI side) creates an auto-dispose reference (will dispose after 3minutes)

**Comment by roblourens** [comment]
So we ref each cell once from the frontend and once from ExtHostNotebooks. I think the issue is that cells are unref'd when the document closes but not when the cell is deleted. So a deleted cell can be leaked to the next time the notebook is opened. Steps

- Create a cell, type in it
- Delete the cell
- Close the notebook
- Reopen the notebook
- Create a cell
- The UI shows an empty cell, but the EH still has the content

![Aug-20-2020 21-52-33](https://user-images.githubusercontent.com/323878/90853606-7ab6f480-e32f-11ea-937e-55aefe88b39a.gif)

**Comment by jrieken** [comment]
Indeed. When deleting a cell `$acceptDocumentAndEditorsDelta` is not called 😱

**Comment by jrieken** [comment]
Oh, I see... This is going through `$acceptModelChanged`
